src/components/TodoItem.tsx

```diff
@@ -0,0 +1,132 @@
+import React, { Dispatch, SetStateAction, useRef, useState } from 'react';
+import { Todo } from '../types/Todo';
+import cn from 'classnames';
+
+type Props = {
+  todo: Todo;
+  isLoading?: boolean;
+  isInEditMode?: boolean;
+  onUpdateTodo: (todo: Todo) => Promise<void>;
+  onRemoveTodo: (todoId: number) => Promise<void>;
+  setEditTodo: Dispatch<SetStateAction<null | number>>;
+};
+
+export const TodoItem: React.FC<Props> = props => {
+  const {
+    todo,
+    isLoading,
+    onRemoveTodo,
+    onUpdateTodo,
+    isInEditMode,
+    setEditTodo,
+  } = props;
+
+  const [newTitle, setNewTitle] = useState(todo.title);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const onSelectTodo = () => {
+    const todoToUpdate = { ...todo, completed: !todo.completed };
+
+    onUpdateTodo(todoToUpdate);
+  };
+
+  const onEditDoubleClick = () => {
+    setEditTodo(todo.id);
+  };
+
+  const onSubmit = async (
+    event: React.FormEvent<HTMLFormElement> | React.FocusEvent<HTMLFormElement>,
+  ) => {
+    event.preventDefault();
+
+    const normalizedTitle = newTitle.trim();
+
+    if (todo.title === normalizedTitle) {
+      setEditTodo(null);
+
+      return;
+    }
+
+    if (isLoading) {
+      return;
+    }
+
+    try {
+      if (normalizedTitle === '') {
+        await onRemoveTodo(todo.id);
+      } else {
+        setNewTitle(normalizedTitle);
+        await onUpdateTodo({ ...todo, title: normalizedTitle });
+      }
+
+      setEditTodo(null);
+    } catch (err) {
+      inputRef?.current?.focus();
+    }
+  };
+
+  const handleKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      setEditTodo(null);
+      setNewTitle(todo.title);
+    }
+  };
+
+  return (
+    <div data-cy="Todo" className={cn('todo', { completed: todo.completed })}>
+      {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={todo.completed}
+          onChange={onSelectTodo}
+        />
+      </label>
+
+      {isInEditMode ? (
```

Comment:
i think - `is` will be enough
```suggestion
      {isEditMode ? (
```